effects: power-up effects analysis for array operations

aviatesk opened the pull request:
Mostly by making use of newly added `:inaccessiblememonly` effect property.
Now we can fold simple vector operations like:
```julia
julia> function simple_vec_ops(T, op!, op, xs...)
           a = T[]
           op!(a, xs...)
           return op(a)
       end;
simple_vec_ops (generic function with 1 method)

julia> for T = Any[Int,Any], op! = Any[push!,pushfirst!], op = Any[length,size],
           xs = Any[(Int,), (Int,Int,)]
           let effects = Base.infer_effects(simple_vec_ops, (Type{T},typeof(op!),typeof(op),xs...))
               @test Core.Compiler.is_foldable(effects)
           end
       end

julia> code_typed() do
           simple_vec_ops(Any, push!, length, Any,nothing,Core.Const(1))
       end
1-element Vector{Any}:
 CodeInfo(
1 ─     return 3
) => Int64
```

aviatesk commented on base/array.jl:
```diff
@@ -1851,7 +1864,7 @@ for (f,_f) in ((:reverse,:_reverse), (:reverse!,:_reverse!))
         $_f(A::AbstractVector, dim::Tuple{Integer}) = $_f(A, first(dim))
         function $_f(A::AbstractVector, dim::Integer)
-            dim == 1 || throw(ArgumentError("invalid dimension $dim ≠ 1"))
+            dim == 1 || _throw_argerror(lazy"invalid dimension $dim ≠ 1")
```
```suggestion
            dim == 1 || _throw_argerror(LazyString(invalid dimension ", dim, " ≠ 1")
```

aviatesk commented:
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/47154#issuecomment-1277678375) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/0e29d79_vs_01310a9/report.md).

vtjnash commented on base/array.jl:
```diff
@@ -396,11 +404,17 @@ julia> getindex(Int8, 1, 2, 3)
     a = Vector{T}(undef, length(vals))
     if vals isa NTuple
         @inbounds for i in 1:length(vals)
-            a[i] = vals[i]
+            a[i] = getfield(vals, i)
```
are you intentionally disabling `inbounds` here, or should this be:
```suggestion
        for i in 1:length(vals)
             a[i] = getfield(false, vals, i)
```

vtjnash commented on base/array.jl:
```diff
@@ -413,9 +427,10 @@ function getindex(::Type{T}, vals...) where T
+    @_safe_getindex_meta
     a = Vector{Any}(undef, length(vals))
     @inbounds for i = 1:length(vals)
-        a[i] = vals[i]
+        a[i] = getfield(vals, i)
```
same here
```suggestion
        for i in 1:length(vals)
             @inbounds(a[i]) = getfield(false, vals, i)
```

vtjnash commented on base/array.jl:
```diff
@@ -1070,11 +1096,12 @@ function push!(a::Vector{Any}, @nospecialize x)
     _growend!(a, nx)
-    for i = 1:nx
-        arrayset(true, a, x[i], na+i)
+    @inbounds for i = 1:nx
```
This does not seem to have any inlining?
```suggestion
    for i = 1:nx
```

vtjnash commented on base/array.jl:
```diff
@@ -1412,11 +1441,12 @@ function pushfirst!(a::Vector{Any}, @nospecialize x)
     _growbeg!(a, nx)
-    for i = 1:nx
-        a[i] = x[i]
+    @inbounds for i = 1:nx
+        a[i] = getfield(x,i)
```
same here

vtjnash commented on base/tuple.jl:
```diff
@@ -65,7 +65,7 @@ end
 function iterate(@nospecialize(t::Tuple), i::Int=1)
     @inline
-    return (1 <= i <= length(t)) ? (@inbounds t[i], i + 1) : nothing
+    return (1 <= i <= length(t)) ? (@inbounds getfield(t,i), i + 1) : nothing
```
```suggestion
    return (1 <= i <= length(t)) ? (getfield(t, i), i + 1) : nothing
```

vtjnash reviewed (commented):
SGTM overall, with questions/comments

aviatesk committed:
effects: power-up effects analysis for array operations

Mostly by making use of newly added `:inaccessiblememonly` effect
property.
Now we can fold simple vector operations like:
```julia
julia> function simple_vec_ops(T, op!, op, xs...)
           a = T[]
           op!(a, xs...)
           return op(a)
       end;
simple_vec_ops (generic function with 1 method)

julia> for T = Any[Int,Any], op! = Any[push!,pushfirst!], op = Any[length,size],
           xs = Any[(Int,), (Int,Int,)]
           let effects = Base.infer_effects(simple_vec_ops, (Type{T},typeof(op!),typeof(op),xs...))
               @test Core.Compiler.is_foldable(effects)
           end
       end

julia> code_typed() do
           simple_vec_ops(Any, push!, length, Any,nothing,Core.Const(1))
       end
1-element Vector{Any}:
 CodeInfo(
1 ─     return 3
) => Int64
```

N5N3 commented on base/array.jl:
```diff
@@ -412,9 +453,21 @@ function getindex(::Type{T}, vals...) where T
     return a
 end
 
+# safe version
+function getindex(::Type{T}, vals::T...) where T
+    @inline
+    @_effect_free_terminates_locally_meta
+    a = Vector{T}(undef, length(vals))
+    @_safeindex for i in 1:length(vals)
+        a[i] = vals[i]
+    end
+    return a
+end
```
@aviatesk
Is there a reason to add this? It seems unneeded and cause some new ambiguity https://github.com/JuliaArrays/StaticArrays.jl/pull/1153#issuecomment-1500790911

aviatesk replied:
I added this definition since otherwise we can't const-fold safe `getindex` operations due to the `:nothrow` effect tainted by the `foldr` branch in the unsafe definition.

aviatesk replied:
But it looks like this definition is no longer needed (the `vals isa NTuple` branch is const-folded even without constant information propagated)? I will make a PR to delete this.

aviatesk replied:
https://github.com/JuliaLang/julia/pull/49310
